miniml/compiler/compile.scm

```diff
@@ -1725,6 +1767,37 @@
          (nenv-modules (add-bindings (env-get-modules env) (env-get-modules menv))))
     (mkenv nenv-vars nenv-constrs nenv-fields nenv-modules)))
 
+(define (mkstruct env defs)
+  (let* ((mark (lambda (e) (vhash-map (lambda (k v) (cons #f (cdr v))) e)))
+         (modenv (mkenv (mark (env-get-vars env))
+                        (mark (env-get-constrs env))
+                        (mark (env-get-fields env))
+                        (mark (env-get-modules env)))))
+    (list 'VModule (compile-defs modenv defs))))
+
+(define (apply-functor mod arg)
+  (assert (equal? (car mod) 'VFunctor))
+  (let* ((fct (car (cdr mod)))
+         (nenv (env-replace-module (functor-get-env fct) (functor-get-argname fct) (cons #f arg))))
+    (match (functor-get-body fct)
+     (('FBase m) (compile-mod nenv m))
+     (('FFunctor argname body) (list 'VFunctor (mkfunctor nenv argname body))))))
+
+(define (compile-functor env args body)
+  (let* ((fct (fold-right (lambda (arg b) (list 'FFunctor arg b)) (list 'FBase body) args)))
+    (match fct
+     (('FBase m) (compile-mod env m))
+     (('FFunctor argname body) (list 'VFunctor (mkfunctor env argname body))))))
```

Comment:
There is a lot of redundancy between the end of this function and the one above.

Reply:
Could we do without the `F` wrapping? If you had a `MEFunctor` as suggested above, you could just use this to build the `fct` module expression (could be called `mod-expr`). Then you call `compile-mod` on it in any case, and that function handles `MEFunctor` by returning a `VFunctor`. How does that sound?